pymc_marketing/mmm/tvp.py

```diff
@@ -78,29 +164,54 @@ def time_varying_prior(
         Regression.
     """
 
-    if X_mid is None:
-        X_mid = float(X.mean().eval())
-    if L is None:
-        L = X_mid * 2
-
     model = pm.modelcontext(None)
 
-    if cov_func is None:
-        eta = pm.Exponential(f"{name}_eta", lam=eta_lam)
-        ls = pm.InverseGamma(f"{name}_ls", mu=ls_mu, sigma=ls_sigma)
-        cov_func = eta**2 * pm.gp.cov.Matern52(1, ls=ls)
+    eta = pm.Exponential(f"{name}_eta", lam=-np.log(eta_mass) / eta_upper)
+
+    if ls_upper is None:
+        ls = pc_prior_1d(f"{name}_ls", alpha=1.0 - ls_mass, lower=ls_lower)
+        ls_upper = 2 * X_mid
+    else:
+        # Note: this function sometimes fails, how to handle?
+        # If not using PC prior, should user set ls_mu and ls_sigma instead?
+        ls_params = pm.find_constrained_prior(
+            pm.InverseGamma,
+            lower=ls_lower,
+            upper=ls_upper,
+            mass=ls_mass,
+            init_guess={"alpha": 2, "beta": 1},
+        )
+        ls = pm.InverseGamma("f{name}_ls", **ls_params)
+
+    lengthscale_range = [ls_lower, ls_upper]
+    m, c = approx_hsgp_hyperparams(
+        X.eval(), X_mid, lengthscale_range, cov_func=cov_func
+    )
+    L = c * X_mid
 
-    model.add_coord("m", np.arange(m))  # type: ignore
+    model.add_coord("m", np.arange(m - 1))  # type: ignore
     hsgp_dims: str | tuple[str, str] = "m"
     if isinstance(dims, tuple):
         hsgp_dims = (dims[1], "m")
 
-    gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
+    cov_funcs = {
+        "expquad": pm.gp.cov.ExpQuad,
+        "matern52": pm.gp.cov.Matern52,
+        "matern32": pm.gp.cov.Matern32,
+    }
+    cov_func = eta**2 * cov_funcs[cov_func.lower()](input_dim=1, ls=ls)
+
+    gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func, drop_first=True)
```

Comment:
shall we pass the `drop_true` from the function signature here?

Reply:
I got hung up on the best default.  There can be identifiability issues between the intercept and a GP, particularly for smaller intercepts (relative to the GP) and at longer lengthscales.  This is evident in the HSGP when c gets a bit larger, the first basis vector becomes more and more constant.  For a default, I think the best option is to either, keep the intercept parameter and set drop_first = True, or have drop_first=False and not have an intercept parameter.  

I'm not sure what would be better though, it depends on what the data "usually" looks like.  Like if the intercept for the TVP is like 1000 and the GP part oscillates between -1 and 1, then there wont be identifiability issues, regardless of lengthscale.  But if the intercept is like 0.2 and the GP oscillates between -1 and 1, then it'll be a problem.